### ISSUE TITLE
Roll over shop cart numbers

### ISSUE THREAD

**Issue by rahaeli** [dreamwidth/dw-free#1898]
Because the shop assigns cart numbers when you load the page, not when you check out an order, the cart numbers increment pretty rapidly. (We're nearing cart 10,000,000, and we definitely haven't had ten million orders...)

We don't need to change this behavior, but since seven-and-eight-digit numbers are hard to copy accurately, we occasionally get check/money order payments with the wrong order # -- digits transposed, digits left out, etc. Let's roll over shop cart numbers somehow so we can get them smaller without having conflicts. 

I originally thought of doing, like, 1-1000, 1-1001, etc, but the more I think about it, the more I think starting with a letter might be better: let's roll over and start with A1000, A1001, A1002, and when we hit A999999, roll over to B1000, B1001, etc.

Parameters:

* Numbers should start with 1,000 (people are weirded by super low numbers) and roll over to the next initial letter at 999,999.

* The initial letters should be capitalized, and exclude I (capital letter i) and O (capital letter o) to avoid font confusion with the numbers 1 and 0.

* It would be super awesome if the places that I look up cart numbers (/admin/pay etc) were case-insensitive, too. :)

**Comment by kareila** [comment]
The problem with changing the format of cart numbers in the database is that there are ten (I counted) different tables that assume cartid is an integer, and that's a LOT of code to audit.

I would instead suggest that we implement a new method, something like `printable_cartid`, which calculates a new format similar to what you describe based on the existing cartid, and use that in the places where the cartid is exposed to the user.

**Comment by rahaeli** [comment]
Yeah, @zorkian said something similar when I was griping to him about this problem. I am all for storing the number as the integer and then doing some pretty transformations to chop the leftmost five numbers (or whatever) off and replace them with a letter and the year or whatever.

**Comment by kareila** [comment]
Since the link has to be reversible in order to lookup the original cartid from the more printable one, the most straightforward approach is probably to create yet another database table to link the two IDs.  That could be done at checkout time, so that only completed carts would have rows in the new table, and the new ID format wouldn't have to be a manipulation of the old one.

> exclude I (capital letter i) and O (capital letter o) to avoid font confusion with the numbers 1 and 0

I think we should also exclude S for the same reason (confusion with the number 5).

**Comment by rahaeli** [comment]
Went to go look for "list of Latin-alphabet letters and numbers that are easily confused", found someone on StackExchange advocating a different solution entirely for activation codes, which is close enough to this that it might work: two short English words, both selected randomly. It would be much easier for me to apply a payment to cart "pretty banana" than cart "19388491043"! 

The thing they linked to was https://wcodes.org/menu which doesn't seem exactly what we want to do (and a 1024-item word list doesn't seem like enough to generate sufficient unique cart #s over time, even if the prettyname would only be assigned at checkout time), but I could come up with something.

If we think that'd be too annoying, the list I found of "easily confused pairs" is B8/G6/I1l/0OQ/S5/Z2
